Fix log-format reporter ignoring install errors

Jordan474 opened the pull request:
When running `spack install --log-format junit|cdash ...`, install errors were ignored. This made spack continue building dependents of failed install, ignoring `--fail-fast`, and exit 0 at the end.

I found it was the `InfoCollector` class not forwarding exceptions after gathering them for report.

**Slight behavior change:** (but for the best) because dependents of a failed install are now correctly skipped (like without `--log-format`), they might not be reported the same as before in the report file (skipped build is not an error).

Fixes #5378

Commit:
Fix log-format reporter ignoring install errors

When running `spack install --log-format junit|cdash ...`, install
errors were ignored. This made spack continue building dependents of
failed install, ignoring `--fail-fast`, and exit 0 at the end.

Fixes #5378

Commit:
Fix install log-format tests

Now that spack install --log-format properly reports failure (just like
when we are not using a --log-format), tests must be fixed.

haampie commented:
Thanks for fixing this 4-digit issue number :sweat_smile: 

Can you also fix the TODO's from the failing tests:

```
        # TODO: Why does junit output capture appear to swallow the exception
        # TODO: as evidenced by the two failing packages getting tagged as
        # TODO: installed?
        with tmpdir.as_cwd():
>           install('--log-format=junit', '--log-file=test.xml', 'libdwarf')
```

bernhardkaindl commented:
@Jordan474 ping for question by haampie:
> Thanks for fixing this 4-digit issue number sweat_smile
> 
> Can you also fix the TODO's from the failing tests:
> 
> ```
>         # TODO: Why does junit output capture appear to swallow the exception
>         # TODO: as evidenced by the two failing packages getting tagged as
>         # TODO: installed?
>         with tmpdir.as_cwd():
> >           install('--log-format=junit', '--log-file=test.xml', 'libdwarf')
```

Jordan474 commented:
@haampie  @bernhardkaindl  Yes, this fixed the `TODO` (I removed the comment): junit does not swallow the exception anymore, and the command fails as expected (I added the assert to test that).

haampie commented:
@spackbot run pipeline

haampie commented:
Rebased it to avoid a lot of redundant recompilation in gitlab
